src/Catalog/Persistence/AzureStorageQueue.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Queue;
+
+namespace NuGet.Services.Metadata.Catalog.Persistence
+{
+    public class AzureStorageQueue<T> : StorageQueue<T>
+    {
+        private CloudQueue _queue;
+
+        private TimeSpan _visibilityTimeout = TimeSpan.FromMinutes(5);
+
+        public AzureStorageQueue(CloudStorageAccount account, string queueName, IStorageQueueMessageSerializer<T> serializer)
+            : base(serializer)
+        {
+            _queue = account.CreateCloudQueueClient().GetQueueReference(queueName);
+            _queue.CreateIfNotExists();
+        }
+
+        public override Task OnAdd(IStorageQueueMessage message, CancellationToken token)
+        {
+            var azureMessage = new AzureStorageQueueMessage(message.Contents);
+            return _queue.AddMessageAsync(azureMessage.Message, token);
+        }
+
+        public override async Task<IStorageQueueMessage> OnGetNext(CancellationToken token)
+        {
+            var nextMessage = await _queue.GetMessageAsync(_visibilityTimeout, null, null, token);
+            if (nextMessage == null)
+            {
+                return null;
+            }
+
+            return new AzureStorageQueueMessage(nextMessage);
+        }
+
+        public override Task OnRemove(IStorageQueueMessage message, CancellationToken token)
+        {
+            return _queue.DeleteMessageAsync((message as AzureStorageQueueMessage).Message, token);
```

Comment:
>AzureStorageQueueMessage [](start = 57, length = 24)

This will throw a hard to debug null exception if the `serializer` returns a `IStorageQueueMessage` that isn't `AzureStorageQueueMessage`. Consider adding a check before the `as`, and if the check fails, throw an exception with a helpful message